internal/abi/fastly/hostcalls_guest.go

```diff
@@ -2480,26 +2484,17 @@ func fastlyDictionaryGet(
 // Get the value for key, as a byte slice, if it exists.
 func (d *Dictionary) GetBytes(key string) ([]byte, error) {
 	keyBuffer := prim.NewReadBufferFromString(key).Wstring()
-	n := DefaultMediumBufLen // Longest (8192) = Config Store limit; typical values likely less than 1024
-	for {
-		buf := prim.NewWriteBuffer(n)
-		status := fastlyDictionaryGet(
-			d.h,
-			keyBuffer.Data, keyBuffer.Len,
-			prim.ToPointer(buf.Char8Pointer()), buf.Cap(),
-			prim.ToPointer(buf.NPointer()),
-		)
-		if status == FastlyStatusBufLen && n < dictionaryValueMaxLen {
-			// The Dictionary API cannot return the needed size with this error.
-			// Instead of perfectly adapting, we allocate the maximum length a value can have.
-			n = dictionaryValueMaxLen
-			continue
-		}
-		if err := status.toError(); err != nil {
-			return nil, err
-		}
-		return buf.AsBytes(), nil
+	buf := prim.NewWriteBufferFromBytes(d.valueBuf[:])
+	status := fastlyDictionaryGet(
+		d.h,
+		keyBuffer.Data, keyBuffer.Len,
+		prim.ToPointer(buf.Char8Pointer()), buf.Cap(),
+		prim.ToPointer(buf.NPointer()),
+	)
+	if err := status.toError(); err != nil {
+		return nil, err
 	}
+	return buf.AsBytes(), nil
```

Comment:
Are our tests sufficient to prove that `buf.AsBytes()` is copying out `valueBuf` to a new buffer, rather than optimizing the slicing out as merely a view inside `valueBuf` itself?

I would imagine something like:

```go
config.Set("foo", "123456789")
config.Set("bar", "abcdefghi")

value1, _ := config.GetBytes("foo")
value2, _ := config.GetBytes("bar")

if value1 != "123456778" { ... }
if value2 != "abcdefghi" { ... }
```

Reply:
You make a good point. I think this is OK as-is:

The slice operation inside `AsBytes` isn't meant to copy, since we're still in `internal` and we control consuming the slice. I wanted to avoid an internal copy to a new `[]byte` before `Get` copies it implicitly with `string()`.

I used the three element form in the shared code so any `append` calls to the result of `AsBytes()` would be forced to make a copy, since that's way more common than directly re-slicing or modifying indices of a `[]byte` in place. This will help protect future additions of shared volatile buffers to other host calls.

The config store code only passes the volatile slice up to the public part of the SDK (to be copied to a new `[]byte`) before end users have access to the data, so perhaps this test is best at that level.

Reply:
This requirement should probably be documented otherwise we're going to forget.  Could we tweak this internal API to make doing the wrong thing harder?

Reply:
> This requirement should probably be documented otherwise we're going to forget. Could we tweak this internal 
API to make doing the wrong thing harder?

@dgryski I agree, this felt risky. Using a mutex is much cleaner and gives a place to document the expected behavior and subtleties.

@lpereira `config.Set` does not exist. 😮